## Introduction {.page_break_before}

*The Introduction should provide context as to why the software tool was developed and what need it addresses. It is good scholarly practice to mention previously developed tools that address similar needs, and why the current tool is needed.*

Substantial growth in publicly available nucleotide sequencing data
(DNA and RNA) has occurred over the last decade, driven by decreases
in sequencing costs. In particular the Sequence Read Archive now has
over XXX PB of data as of YY date, with YYY PB of data public.  Given
the sheer volume of data, finding data sets relevant to a particular
scientific question among the many publicly available data is
difficult.

Shotgun metagenomes are a particularly intriguing type of public data.
These data sets are generated by random sequencing of mixtures of
microbes sampled from a microbiome. Metagenome data sets are often
large (100s of MBs to 10s of GB) and can be highly complex, with
environmental samples containing genomic data that can be attributed
to thousands or more species. In many cases these metagenomes come
from microbiomes containing many organisms unknown at the time of
sequencing, and a productive cottage industry of data reusers has
emerged. In the past decade, hundreds of thousands of new bacterial
and archaeal genomes have been isolated from public metagenomes, and
several entirely new branches of life have been discovered (large
bacteriophage, hug et al, etc.)

There are many use cases for SRA-scale search that rely on
content-based search (see Table XX). This is in part because of the
many unknowns present in metagenomes, and the associated unreliability
of annotations generated at the time of submission; but bioinformatics
data analysis techniques such as de novo assembly are also typically
lossy, and so reanalyzing data with different parameters is often
fruitful. In particular, a number of studies have shown that it's a
good idea to search unassembled data rather than building MAGs first.

However, content-based search of 100s of thousands of metagenomes is
technically challenging.  This is largely because of their sheer size
and diversity; approaches that work for single-organism genomic and
RNAseq data sets such as SBTs and BIGSI do not match well to datasets
with unknown levels of underlying sequence diversity, because they
must choose a single parameter for their Bloom filters. Serratus
requires access to the raw metagenome data, which can be costly.

Below, we introduce MAGsearch, an SRA-scale querying system that uses
containment searches based on FracMinHash sketching to search all
public metagenome data sets in the SRA in approximately 24 hours on
commodity hardware with 1-1000 query genomes.  MAGsearch uses the Rust
library underlying the sourmash implementation of FracMinHash to
execute massively parallel searches of a presketched digest of the
SRA. This addresses many (but not all) of the use cases in Table XXX.

The availability of relatively lightweight content-based search of SRA
metagenomes presents a possible solution for a number of use cases (Table XX).
Some of these use cases have already been explored with MAGsearch:
Viehweger et al. used MAGsearch to discover a metagenomic sample containing
*Klebsiella pneumonia* that was subsequently included in an outbreak
analysis, and Lumian et al. (2022) conducted a biogeographical study
on five newly generated cyanobacterial genomes from Antarctic samples.


Thoughts and questions:

* do we mention CMash and mash screen? Same challenges as serratus and
  intended for single sample analysis.
  not intended for large scale search but rather s
* (re)emphsaize unassembled, no marker genes
* figure out how to talk about things it can't do - in Table 1, and
  also caption, and also discussion.
* For discussion: content-based search, including MAGs and things
  without marker genes.
* stress lightweight resource usage

----

Table 1 contents to have:

* Biogeography of genomes
  * describe and characterize biogeographical distribution of species and genus;
  * identify potential sampling locations;
  * (Jessica paper)
* Outbreak tracking
  * trace history of pathogen evolution from public data
  * monitor for future outbreaks
  * (Adrian paper)
* Expanding / exploring strain, species, and genus composition for further genomic analysis
  * Newly isolated species and genera
  * SAGs
* Content-based exploration and reannotation of SRA contents vs metadata
  * Host contamination
  * IBD stuff/discovering gut microbiome
* Making other large-scale databases accessible
  * differential privacy
  * commercial privacy
* Post-processing and cleaning MAGs
* Discovering matches in newly public samples; notification service
* Exploring spread of AMR
* Regulatory evaluation of probiotics

Small viral pangenome query comment/Luiz. Association studies?

Anti use cases: small virus search a la serratus. Searching for new
members of class or order. Searching for specific functional genes.
