## Introduction {.page_break_before}

*The Introduction should provide context as to why the software tool was developed and what need it addresses. It is good scholarly practice to mention previously developed tools that address similar needs, and why the current tool is needed.*

Substantial growth in publicly available nucleotide sequencing data
(DNA and RNA) has occurred over the last decade, driven by decreases
in sequencing costs. In particular the Sequence Read Archive now has
over XXX PB of data as of YY date.

A particularly intriguing sort of data is metagenomes, shotgun
(random) sequencing of mixtures of microbes from host-associated and
environmental samples. These datasets are often large, and can be
highly complex, with environmental samples containing genomic sequence
from thousands or more species. These samples often contain many
microbial species unknown at the time of their deposition as well. The
SRA currently contains about 800,000 of these data sets as of ZZ date,
and these datasets are an as-yet largely untapped source of
information on microbial life. In particular, they are ripe for reuse,
since they typically contain culture-independent random samples of an
entire microbial community, even when they were collected for a
specific purpose.

Mining SRA metagenomes for biological information takes many forms,
including reference-based and de novo techniques. De novo techniques
focus on isolating new genomes of certain types from large
collections. Reference based techniques concentrate instead on
identifying new strains and species organisms related to known species
using genome references; this has been used on a large scale with
serratus. However, all of these techniques are relatively heavyweight
and in particular rely on analyzing the original data, which is time
consuming and expensive. The diversity of these data prevent
approaches developed for single-organism data such as BIGSI from being
used. (Also mention raptor? Or others?)

The growth in k-mer sketching approaches that support containment
analysis, i.e. looking for small overlaps between two sequencing data
sets, provides an opportunity to enable fast queries on a much reduced
index of the original data. In particular, the FracMinHash approach
supports robust overlap and containment analysis on sketches, and the
sourmash software implementation provides a mature CLI and Python/Rust
library for generating and investigating FracMinHash sketches.

Here, we describe a fast search tool, MAGsearch, that makes use of
precomputed sourmash FracMinHash sketches to search the entire
metagenomic SRA in under 24 hours in 50 GB of RAM and 32 threads,
i.e. on commodity hardware, 1-1000 microbial genomes at a time. With
the present SRA digest, this tool has high sensitivity and specificity
for detecting 10kb of larger overlaps, and can also be used to detect
larger matches at average nucleotide identities of 90% or lower,
i.e. genus level matches.

We developed a new tool for the specific use case of searching the SRA
because the general purpose sourmash command-line tool and Python
library was not performant enough to search.

### Biological use cases for MAGsearch

The availability of relatively lightweight search of SRA metagenomes
has opened up a number of use cases, including biogeographical
analysis of genomes and pathogen outbreak analysis. And, because
sketches can be generated separately and this approach searches only a
subset of the information that does not require revealing the full
sample data, it can also be applied to making private metagenome
databases available for search, e.g. for commercial collections of
metagenomes.

features of current MAGsearch:
* content-based search, including MAGs and things without marker genes.
* lightweight resource usage

* Biogeography of genomes
  * describe and characterize biogeographical distribution of species and genus;
  * identify potential sampling locations;
  * (Jessica paper)
* Outbreak tracking
  * trace history of pathogen evolution from public data
  * monitor for future outbreaks
  * (Adrian paper)
* Expanding / exploring strain, species, and genus composition for further genomic analysis
  * Newly isolated species and genera
  * SAGs
* Content-based exploration and reannotation of SRA contents vs metadata
  * Host contamination
  * IBD stuff/discovering gut microbiome
* Making other large-scale databases accessible
  * differential privacy
  * commercial privacy
* Post-processing and cleaning MAGs
* Discovering matches in newly public samples; notification service
* Exploring spread of AMR
* Regulatory evaluation of probiotics

Small viral pangenome query comment/Luiz. Association studies?

Anti use cases: small virus search a la serratus. Searching for new
members of class or order. Searching for specific functional genes.
