## Introduction {.page_break_before}

<!-- *The Introduction should provide context as to why the software tool was developed and what need it addresses. It is good scholarly practice to mention previously developed tools that address similar needs, and why the current tool is needed.* -->

Substantial growth in publicly available nucleotide sequencing data
(DNA and RNA) has occurred over the last decade, driven by decreases
in sequencing costs.  In particular the Sequence Read Archive now has
over 9 million entries containing 12 PB of data [@sra].  Shotgun
metagenomes, generated by random sequencing of mixtures of microbes
sampled from a microbiome are a particularly interesting resource
stored in the SRA.

Shotgun metagenome data sets are often large (100s of MBs to 10s of
GB) and can be highly complex, with environmental samples containing
genomic data that can be attributed to thousands or more species.  In
the past decade, hundreds of thousands of new bacterial and archaeal
genomes have been isolated from public metagenomes, and several
entirely new branches of life have been discovered purely through
analysis of public data
[@doi:10.1038/s41587-020-0603-3; @doi:10.1038/nmicrobiol.2016.48; @doi:10.1038/s41586-020-2007-4].

Beyond their initial use, these data sets form an incredibly rich
resource for contextualizing novel sequencing data and for synthesis
research on a myriad of large-scale genomic questions ranging from
basic evolutionary processes to disease associations and pathogenicity
tracking (Table @tbl:bio-use-cases).
However, comprehensive discovery of relevant data sets is challenging.
Metadata for these data sets is typically geared towards the
submitting researcher's study questions and major findings, and moreover cannot
possibly describe the full contents of the data.
Furthermore, metadata provided at the time of submission can be
incomplete or inconsistent, rendering systematic data set discovery
intractable.

Content-based search is a promising alternative strategy for finding
data sets in archives.  By searching with genomic content of interest,
content-based search can recover datasets containing relevant species
or genes of interest regardless of their associated metadata.
However, search of unassembled sequence is critical to ensure unbiased
and comprehensive recovery of relevant datasets.  Assembly techniques
are designed to produce consensus reference sequences useful for
consistent comparisons across genotypes, often collapsing sequence
variation in the process [@doi:10.1038/nmeth1043;@doi:10.1101/155358].
In addition, reassembly and reanalysis of existing data using
different parameters or newer methods often yields different results.
Content-based search of unassembled metagenomes can bypass these
issues and facilitate consistent downstream analysis across data sets
that may have been initially generated to answer a range of disparate
biological questions, and been first analyzed over a range of years
and with myriad techniques.

A number of approaches have been developed to enable content-based
search of single-organism genomic and RNAseq data.  Methods that
enable rapid, large-scale search across hundreds of thousands of data
sets typically leverage biological sketching techniques and
probabilistic data structures to reduce the effective search space
[@sbt; @bigsi; @ssbt].  However, these approaches do not readily
translate to datasets with unknown levels of sequence diversity, the
defining feature of metagenomic datasets.

Recently Serratus used extensive search across public datasets to
recover 880,000 RNA-dependent RNA polymerase-containing sequences, and
discovered over 131,000 novel RNA viruses [@serratus].  This search
was comprehensive but also time-consuming and costly, and still out of
reach for unfunded exploratory research. Other approaches such as
searchsra.org [@searchsra] and metagraph [@metagraph] are promising
but are not yet capable of searching all public data.

<!-- add https://www.biorxiv.org/content/10.1101/2022.10.07.510278v1.full at some point -->

Here, we introduce Branchwater, a petabase-scale querying system that
uses containment searches based on FracMinHash sketching to search all
public metagenome data sets in the SRA in 24-36 hours on
commodity hardware with 1-1000 query genomes. Branchwater uses the
Rust library underlying the sourmash implementation of FracMinHash to
execute massively parallel searches of a presketched digest of the
SRA [@sourmash_joss; @gather].

The availability of relatively lightweight content-based search of SRA
metagenomes helps address many of the biological use cases in Table
@tbl:bio-use-cases (see 3rd column).  Some of these use cases have already been
explored with Branchwater: Viehweger et al. (2021) [@viehweger] used
Branchwater to discover a metagenomic sample containing *Klebsiella
pneumonia* that was subsequently included in an outbreak analysis, and
Lumian et al. (2022) [@lumian_biogeo] conducted a biogeographical
study on five newly generated cyanobacterial metagenome-assembled genomes from Antarctic
samples.

| Use case | Description | Enabled by branchwater |
| -------- | -------- | -------- |
| Biogeography of genomes     | Describe and characterize biogeographical distribution of species; identify sampling locations     | Yes     |
| Outbreak tracking | Trace pathogen spread via public data | Yes, for genomes > 10kb |
| Pangenome expansion | Expand and explore composition of strains, species and genus level pangenomes (including SAGs and MAGs) | Yes |
| SRA metadata reannotation | Content-based validation and reannotation of SRA metagenomes | Yes |
| Private database access and search | Provide privacy-enabled search of large, access-restricted databases | Yes |
| Post-processing and cleaning MAGs | Evaluating contig-level presence and abundance across multiple metagenomes | Yes, for contigs > 10kb |
| Exploring breadth of plasmids etc. | Evaluating range and prevalence of laterally transferred genomic elements | Yes, for plasmids > 10kb |
| Exploring host range of species for regulatory evaluation | | Yes, for genomes > 10kb |
| Search for small viruses | | No |
| Searching for specific functional genes | | No |
| Detecting novel classes and orders | |  No |

Table: Biological use cases for petabase scale sequence search of metagenomes {#tbl:bio-use-cases}

<!-- edit table [here](https://hackmd.io/Pgj6AjM_RlGF0_7rGhY3tA) -->

<!--

Thoughts and questions:

* For discussion: content-based search, including MAGs and things
  without marker genes.
* stress lightweight resource usage
* note that a key requirement of Branchwater was that we search unassembled.
* discuss thresholds ...somewhere.

CTB NOTE: Small viral pangenome query comment/Luiz. Association studies.

-->
