## Background: FracMinHash and sourmash {.page_break_before}

FracMinHash is a bottom-sketch version of ModHash that supports
accurate estimation of overlap and containment between two sequencing
sets (gather paper). In brief, FracMinHash is a lossy compression
approach that reduces data sets in size by a scaled factor S. Sketches
support estimation of overlap, bidirectional containment, and Jaccard
similarity between two data sets. Unlike other common sketching
techniques (MinHash, HyperLogLog), FracMinHash supports these
operations between two data sets of different sizes, and unlike mash
screen and CMash does not require the original data sets.

The open-source sourmash software provides a mature and
well-documented command-line interface to FracMinHash, along with
Python and Rust APIs for loading and using FracMinHash sketches. The
Python layer provides a larger number of user experience conveniences
on top of the performant Rust layer. However, despite the thread
safety of the underlying Rust code, the CLI and Python library still
operate in single-threaded mode, which limits the utility of sourmash
for very large scale operations. Refactoring the sourmash CLI and
Python libraries to take advantage of thread safety is a substantial
and ongoing effort and we chose instead to develop a dedicated CLI
in Rust.

There are several limitations of FracMinHash and sourmash that limit
their utility to specific downstream use cases. In particular, the
default scaled parameters used below do not work well for genomes
smaller than 10kb in size. Nor can divergent genomes be found; based
on the k-mer containment to ANI conversion, we find that sourmash and
Branchwater work well for finding matches to genomes within about 10%
ANI of the query. Finally, FracMinHash was developed for shotgun data
sets and different parameters would be required for targeted
sequencing data such as amplicon data sets.  Some of these limitations
are intrinsic to FracMinHash, and others may be overcome by parameter
tuning and further research.

### Petabase scale search represents a specific technical challenge to sourmash

The primary design focus for the sourmash CLI has been on searching
and comparing many genome-sized sketches, where for
typical parameters there are between 1000 and 10,000 hashes in each sketch. The
software provides a variety of in-memory and on-disk data structures
for organizing sketches in this size range and can search hundreds of
thousands of genome sketches with a single query in minutes in a
single thread on an SSD laptop; more complex algorithms such as the
min-set-cov described in XXX can take a few hours but are still
acceptably performant on real-world data.

Branchwater faces very different parameters in searching 800,000
metagenomes.  Many of these data sets are extremely large, slow to
read from disk, and individually require substantial memory to load.
Multiple queries may be used to search each metagenome as well, making
this a quadratic search.

One solution we tried initially was a scatter-gather approach based on
a cluster-aware workflow engine (in this case, snakemake). The
overhead on workflow coordination and executing shell commands was
prohibitive, so we implemented a purpose-built multithreaded solution
instead.
