## Background: FracMinHash and sourmash {.page_break_before}

FracMinHash is a bottom-sketch version of ModHash that supports
accurate estimation of overlap and containment between two sequencing
sets [@gather]. In brief, FracMinHash is a lossy compression approach
that represents data sets using a "fractional" sketch containing 1/S
of the original k-mers.  Sketches support estimation of overlap,
bidirectional containment, and Jaccard similarity between two data
sets. Unlike other common sketching techniques such as MinHash [@mash]
and HyperLogLog [@dashing], FracMinHash supports these operations
between two data sets of different sizes, and unlike mash screen and
CMash does not require the original data sets [@mash_screen; @cmash].
In exchange, FracMinHash sketches are essentially unbounded in size.

The open-source sourmash software provides a mature and
well-documented command-line interface to FracMinHash, along with
Python and Rust APIs for loading and using FracMinHash sketches
[@sourmash_joss; @sourmash_f1000]. The Python layer provides a larger
number of user experience conveniences on top of the performant Rust
layer. However, despite the thread safety of the underlying Rust code,
the CLI and Python library still operate in single-threaded mode,
which limits the utility of sourmash for very large scale
operations. Refactoring the sourmash CLI and Python libraries to take
advantage of thread safety is a substantial and ongoing effort; we
chose here to develop a dedicated CLI in Rust instead.

There are several features of FracMinHash and sourmash that limit
their utility for specific use cases. In particular, the default
scaled parameters used in sourmash do not work well for comparing or
detecting genomes smaller than 10kb in size. Nor can divergent genomes
be found; based on k-mer containment to ANI conversion
[@dk_fracminhash], we find that sourmash works well for finding
matches to genomes within about 90% ANI of the query, but not
further. Finally, FracMinHash was developed for shotgun data sets and
different parameters would be required for targeted sequencing data
such as amplicon data sets.  Some of these limitations are intrinsic
to FracMinHash, and others may be overcome in the future by parameter
tuning and further research.

### Petabase scale search represents a specific technical challenge to sourmash

The primary design focus for the sourmash CLI has been on searching
and comparing many genome-sized sketches, where for
typical parameters there are between 1000 and 10,000 hashes in each sketch. The
software provides a variety of in-memory and on-disk data structures
for organizing sketches in this size range and can search hundreds of
thousands of genome sketches with a single query in minutes in a
single thread on an SSD laptop; more complex algorithms such as the
min-set-cov described in XXX can take a few hours but are still
acceptably performant on real-world data.

Branchwater faces very different parameters in searching 800,000
metagenomes.  Many of these data sets are extremely large, slow to
read from disk, and individually require substantial memory to load.
Multiple queries may be used to search each metagenome as well, making
this a quadratic search.

One solution we tried initially was a scatter-gather approach based on
a cluster-aware workflow engine (in this case, snakemake
[@snakemake]). The overhead on workflow coordination and executing
shell commands was prohibitive for our initial implementation, so we
pursued a purpose-built multithreaded solution instead.
