## Results {.page_break_before}

### Example Branchwater search results

To showcase the utility of Branchwater search, we searched a gut bacterium
mixture against all indexed SRA metagenomes using a k-mer size of 31.
Branchwater search is exhaustive: every query is searched against every
search metagenome. Reporting resources are minimal: each match is reported
as a single line in a CSV containing the query and match identities and the
k-mer containment of the query in the match. As a result, we use a low default
threshold, 0.01, which means that any metagenome that contains more than 1% of
the k-mers in any query is reported.
Branchwater search with this default threshold returned 192,699 metagenomes.
We then filtered these results for metagenomes that contained at least 20% of
query k-mers, retaining 66,705 metagenomes for further analysis.
The summarized SRA metadata (Table @tbl:match_categories) provides some insight into
the types of metagenomes recovered: the majority were annotated as "human gut
metagenome" or "gut metagenome", with a smaller number of other gut-related
categories, including "wastewater metagenome" and "mouse gut metagenome".
Branchwater also reported 167 "sediment metagenome" samples, but
upon further investigation, many of these originated from a transect study
investigating the presence of microbes with distance from wastewater treatment.

| ScientificName         |   count of matches |
|:-----------------------|-----------------:|
| human gut metagenome   |            41089 |
| gut metagenome         |             6933 |
| metagenome             |             6144 |
| human metagenome       |             4666 |
| Homo sapiens           |             3141 |
| feces metagenome       |             2261 |
| wastewater metagenome  |             1279 |
| mouse gut metagenome   |              191 |
| human feces metagenome |              184 |
| sediment metagenome    |              167 |

Table: Output summary of `ScientificNames` from metagenome annotations in the Sequence Read Archive for a gut bacterium search mixture. We have provided a simple script that imports the SRA metadata and summarizes the Branchwater results at the provided threshold. {#tbl:match_categories}

### Performance and scaling analysis

In Tables @tbl:avg_bench and @tbl:catalog we show performance metrics
for branchwater. Table @tbl:avg_bench contains average measurements
and standard deviations for time, memory, and I/O, showing that
branchwater is I/O and memory intensive. Table @tbl:catalog compares
the runtimes and memory usage for a search the entire catalog
(normalized to 10k samples) against both the random subsets in Table
@tbl:avg_bench and the largest 10k sketches in the database. The
increased memory usage of the catalog and especially the 10k largest
metagenomes suggests that a relatively small number of metagenomes
contributes the bulk of both time and memory usage to a full catalog
search of branchwater.

| metric | observed | 
| -------- | -------- | 
| time     | 24.2 +/- 1.7 min |
| max RSS | 16.4 +/-  1.6 GB |
| I/O in | 93.4 +/- 1.9 GB |

Table: Time, memory, and I/O input for 5 runs of 1000 queries against 10,000 metagenomes. Queries were randomly selected from 318k genomes in GTDB rs207. Metagenomes were randomly selected from the full catalog of 767k. {#tbl:avg_bench}

| database | time | max RSS |
| -------- | -------- | -------- |
| entire catalog | 27.2m$^*$ | 52.5 GB |
| random 10k subset | 24.2m   | 16.4 GB |
| 10k largest metagenomes | 594.0 m | 60.0 GB |

Table: A comparison of database searches for the entire catalog of
767k metagenomes, with time normalized to 10k, the average of the 10k
subsets from Table @tbl:avg_bench, and the 10k largest metagenomes. {#tbl:catalog}

![**branchwater scales well with number of threads.** Processing time drops linearly with number of threads, while total compute stays approximately the same and memory usage increases linearly with number of threads as each thread loads a subject to search.](images/basic_benchmarks.svg "basic benchmarking"){#fig:basic_bench}

<!-- CTB: should note that one thread is used exclusively for writing to CSV. -->

![**branchwater scales linearly with number of queries and subjects, but number of subjects dominates runtime.** Processing time increases slowly with number of query genomes used to search, because they are held in memory and fast to compare. Processing time increases quickly with number of subject metagenomes being searched, because they are large and slow to load and search.](images/subsample_benchmarks.svg "branchwater scaling"){#fig:basic_scaling}

See https://github.com/dib-lab/2022-branchwater-benchmarking for
number processing and figure generating notebooks.

### Post-search validation

The sourmash CLI can be used to explore k-mer matches for individual
data sets. This does not validate the matches beyond confirming the
containment numbers, although sourmash provides additional information
(e.g. estimated abundances) on top of the minimal information
provided by branchwater.  FracMinHash generally and Branchwater
specifically have been validated bioinformatically with read mapping
(see [@gather; @lumian_biogeo]). This is discussed
further below.

<!-- talk about detection more... ->

<!--
### branchwater is inexpensive and supports exploratory queries

CTB: Estimate cost of a run. Compare to serratus - cloud compute, data
download. Serratus is probably cheaper than $20k now but still
expensive.

-->
